### PR TITLE
Adding positional superko check

### PR DESCRIPTION
**Pull request by robert-waite**

### CODE REVIEW

File: AlphaGo/go.py [hunk 1]
```diff
@@ -54,6 +61,15 @@ def __init__(self, size=19, komi=7.5):
 		# on-the-fly record of 'age' of each stone
 		self.stone_ages = np.zeros((size, size), dtype=np.int) - 1
 
+		# setup Zobrist hash of board state
+		self.enforce_superko = enforce_superko
+		self.current_hash = np.uint64(0)
+		for i in range(0, size):
```
Comment (wrongu) [inline AlphaGo/go.py]: `for x in range(size)`

(`i` and `j` are not used?)

File: AlphaGo/go.py [hunk 2]
```diff
@@ -167,6 +185,10 @@ def _remove_group(self, group):
 					for (gx, gy) in self.group_sets[nx][ny]:
 						self.liberty_counts[gx][gy] = len(self.liberty_sets[nx][ny])
 
+	def _update_hash(self, x, y, from_color, to_color):
```
Comment (wrongu) [inline AlphaGo/go.py]: nit: the convention in this file is to pass `(x,y)` around as a tuple.

File: AlphaGo/go.py [hunk 3]
```diff
@@ -14,19 +14,26 @@ class GameState(object):
 	# amount of time, hence this shared lookup table {boardsize: {position: [neighbors]}}
 	__NEIGHBORS_CACHE = {}
 
-	def __init__(self, size=19, komi=7.5):
+	# Zobrist hash values for each possible state for a given position
+	# Using 25 as max expected board size
+	__HASH_LOOKUP = {
+		EMPTY: np.random.randint(0, np.iinfo(np.uint64).max, size=(25, 25), dtype='uint64'),
```
Comment (wrongu) [inline AlphaGo/go.py]: for the sake of consistency, we might want to set numpy's random seed at the top of the file.

also, why use `25` when you could make it the size of the actual board (as an instance variable)? Or, you could follow the convention in `__NEIGHBORS_CACHE` and create a class-level variable that itself can handle any board size.

File: AlphaGo/go.py [hunk 4]
```diff
@@ -14,19 +14,26 @@ class GameState(object):
...
+	def __init__(self, size=19, komi=7.5, enforce_superko=False):
 		self.board = np.zeros((size, size))
 		self.board.fill(EMPTY)
 		self.size = size
 		self.turns_played = 0
 		self.current_player = BLACK
 		self.ko = None
-		self.komi = komi
+		self.komi = komi  # Komi is number of extra points WHITE gets for going 2nd
+		self.handicaps = []
```
Comment (wrongu) [inline AlphaGo/go.py]: handicaps are from another PR, right? (one that we can merge soon, hopefully!)

File: AlphaGo/go.py [hunk 5]
```diff
@@ -150,7 +166,9 @@ def _remove_group(self, group):
 		updating group sets and liberties along the way
 		"""
 		for (x, y) in group:
-			self.board[x, y] = EMPTY
+			if self.board[x, y] != EMPTY:
```
Comment (wrongu) [inline AlphaGo/go.py]: There *shouldn't* ever be a case where `group` contains `EMPTY` positions.

File: AlphaGo/go.py [hunk 6]
```diff
@@ -212,18 +237,56 @@ def is_suicide(self, action):
 			return True
 		return False
 
+	def is_positional_superko(self, action):
+		if len(self.handicaps) == 0 and self.current_player == BLACK:
+			player_history = self.history[0::2]
+		elif len(self.handicaps) > 0 and self.current_player == WHITE:
+			player_history = self.history[0::2]
+		else:
```
Comment (wrongu) [inline AlphaGo/go.py]: so.. I see why you wanted to include handicaps in this PR. Let's make sure the other one is merged first. Then, please add a comment or two explaining these lines, e.g. "Find all actions that the current_player has done in the past, taking into account the fact that history starts with BLACK when there are no handicaps or with WHITE when there are."

File: AlphaGo/go.py [hunk 7]
```diff
@@ -212,18 +237,56 @@ def is_suicide(self, action):
...
+			player_history = self.history[1::2]
+
+		if action not in self.handicaps and action not in player_history:
+			return False
+
+		color = self.current_player
+		state_copy = self.copy()
+
+		(x, y) = action
+		state_copy.board[x][y] = color
```
Comment (wrongu) [inline AlphaGo/go.py]: Is there an advantage to doing this over `state_copy.do_move(action)`?

**Comment by wrongu** [comment]
This is a great contribution. Could you also add some tests? One showing that it detects superko (ideally demonstrating this where a capture is included), one showing that this does/doesn't affect the value of `is_legal` when the check is enabled/disabled.

File: AlphaGo/go.py (hunk 1 above)
Comment (robert-waite) [inline AlphaGo/go.py]: yeah... thats awful

File: AlphaGo/go.py (hunk 5 above)
Comment (robert-waite) [inline AlphaGo/go.py]: So... I didn't want to spend the time to fully understand how remove group works. I think when I was trying to get the hashing to work... I felt that it was possible that a coordinate might be marked in two groups... seems doubtful at the moment saying that... but wanted a check to make sure that I didn't rehash a removed stone. Rehashing a coordinate has the effect of removing it.

File: AlphaGo/go.py (hunk 7 above)
Comment (robert-waite) [inline AlphaGo/go.py]: do_move calls is_legal... so didn't want an infinite loop. In general... I am attempting to impact your code as little as possible. Felt like a flag to a key method like do_move seemed ugly. Or just didn't refactor enough before pr. Additionally... I think this is the minimal code needed to update the hash and complete a move with captures. A future dev would have less to wade through by seeing it and could possibly figure out rolling back a play.

File: AlphaGo/go.py (hunk 3 above)
Comment (robert-waite) [inline AlphaGo/go.py]: I didn't want to create it every time. It seems like python accepts what I did as a static var of sorts? And really we just need a lookup of random values that is consistent for an instance of the application. So figure there surely is an upper limit to board size. Is a 100x100 really something anyone is interested in?

As for random seed... did think of that.. but when putting together was heading for functionality first.. and didn't wanna read too deeply on the rng of numpy. My main concern at the time was that the ints being created were at least 64 bits... and not messed up somehow between numpy c libs and python.

File: AlphaGo/go.py (hunk 7 above)
Comment (wrongu) [inline AlphaGo/go.py]: I'd argue that re-using `do_move` would not "impact" other code so much as "leverage" it. Agreed that refactoring these set-updating functions would help in the future. In the meantime, reusing things makes them easier to maintain.

What if you set `state_copy.enforce_superko = False` before calling `state_copy.do_move`?

File: AlphaGo/go.py (hunk 3 above)
Comment (robert-waite) [inline AlphaGo/go.py]: I think when i was reading through... felt that the way __NEIGHBORS_CACHE was built didn't look very threadsafe. Although... i hadnt played too much with python before... and guess threading isnt quite the same as other languages.

**Comment by robert-waite** [comment]
Not sure how the flake8 error suddenly showed. Flake8 local doesnt show this

**Pull request merged**